twitter_scrooge/twitter_scrooge.bzl

```diff
@@ -0,0 +1,269 @@
+
+  only_transitive_thrift_srcs_file = ctx.new_file(ctx.outputs.srcjar, ctx.outputs.srcjar.short_path + "_only_transitive_thrift_srcs")
+  ctx.file_action(output = only_transitive_thrift_srcs_file, content = _content_newline(only_transitive_thrift_srcs))
+
+  immediate_thrift_srcs_file = ctx.new_file(ctx.outputs.srcjar, ctx.outputs.srcjar.short_path + "_immediate_thrift_srcs")
+  ctx.file_action(output = immediate_thrift_srcs_file, content = _content_newline(immediate_thrift_srcs))
+
+  ctx.action(
+    #TODO do the dependencies of the executable need to be listed?
+    executable = ctx.executable._pluck_scrooge_scala,
+    inputs = list(remote_jars) +
+        list(only_transitive_thrift_srcs) +
+        list(immediate_thrift_srcs) +
+        cjars +
+        [remote_jars_file,
+         only_transitive_thrift_srcs_file,
+         immediate_thrift_srcs_file],
+    outputs = [ctx.outputs.srcjar],
+    arguments = [
+      only_transitive_thrift_srcs_file.path,
+      immediate_thrift_srcs_file.path,
+      ctx.outputs.srcjar.path,
+      remote_jars_file.path,
+    ],
+    progress_message = "creating scrooge files %s" % ctx.label,
+  )
+
+  jars = _collect_scalaattr(ctx.attr.deps) # should be _collect_scalaattr
+
+  scalaattr = struct(outputs = None,
+                     transitive_runtime_deps = jars.transitive_runtime_deps, #TODO are we missing any runtime deps?
+                     transitive_compile_exports = jars.transitive_compile_exports + set(transitive_cjars),
+                     transitive_runtime_exports = jars.transitive_runtime_exports,
+                     )
+
+  transitive_srcjars = collect_srcjars(ctx.attr.deps) + collect_extra_srcjars(ctx.attr.deps)
+
+  srcjarsattr = struct(
+    srcjar = ctx.outputs.srcjar,
+    transitive_srcjars = transitive_srcjars,
+  )
+
+  return struct(
+    scala = scalaattr,
+    srcjars=srcjarsattr,
+    extra_information=[struct(
+      srcjars=srcjarsattr,
+      scrooge_srcjar=struct(transitive_owned_srcs = transitive_owned_srcs + immediate_thrift_srcs),
+    )],
+  )
+
+def _collect_scalaattr(targets):
+  transitive_runtime_deps = set()
+  transitive_compile_exports = set()
+  transitive_runtime_exports = set()
+  for target in targets:
+    if hasattr(target, "scala"):
+      transitive_runtime_deps += target.scala.transitive_runtime_deps
+      transitive_compile_exports += target.scala.transitive_compile_exports
+      if hasattr(target.scala.outputs, "ijar"):
+        transitive_compile_exports += [target.scala.outputs.ijar]
+      transitive_runtime_exports += target.scala.transitive_runtime_exports
+
+  return struct(
+    transitive_runtime_deps = transitive_runtime_deps,
+    transitive_compile_exports = transitive_compile_exports,
+    transitive_runtime_exports = transitive_runtime_exports,
+  )
+
+scrooge_scala_srcjar = rule(
+    _gen_scrooge_srcjar_impl,
+    attrs={
+        "deps": attr.label_list(mandatory=True),
+        #TODO we should think more about how we want to deal
```

Comment:
these are jars that have both thrift and scala code in them, right?

Can we exercise this in tests?